src/Resource.php

```diff
@@ -0,0 +1,153 @@
+<?php
+/**
+ * Part of the Joomla Framework DI Package
+ *
+ * @copyright  Copyright (C) 2013 - 2014 Open Source Matters, Inc. All rights reserved.
+ * @license    GNU General Public License version 2 or later; see LICENSE
+ */
+
+namespace Joomla\DI;
+
+/**
+ * Defines the representation of a resource.
+ *
+ * @since  2.0
+ */
+class Resource
+{
+	const NO_SHARE = 0;
+	const SHARE = 1;
+	const NO_PROTECT = 0;
+	const PROTECT = 2;
+
+	/** @var Container */
+	private $container;
+
+	/** @var  mixed */
+	private $instance = null;
+
+	/** @var \Callable */
+	private $factory = null;
+
+	/** @var bool */
+	private $shared = false;
+
+	/** @var bool */
+	private $protected = false;
+
+	/**
+	 * Create a resource representation
+	 *
+	 * @param   Container  $container  The container
+	 * @param   mixed      $value      The resource or its factory closure
+	 * @param   int        $mode       Resource mode, defaults to Resource::NO_SHARE | Resource::NO_PROTECT
+	 */
+	public function __construct(Container $container, $value, $mode = 0)
```

Comment:
Correctly, the parameter should be `self::NO_SHARE | self::NO_PROTECT`. But that does not work (see build #58), so I chose the literal value over introducing a new constant (or make it looking/feeling) incomplete.